google/cloud/storage/internal/service_account_credentials.h

```diff
@@ -0,0 +1,176 @@
+
+#include "google/cloud/storage/credentials.h"
+#include "google/cloud/storage/internal/credential_constants.h"
+#include "google/cloud/storage/internal/curl_request.h"
+#include "google/cloud/storage/internal/nljson.h"
+#include "google/cloud/storage/internal/openssl_util.h"
+#include <chrono>
+#include <condition_variable>
+#include <ctime>
+#include <iostream>
+#include <mutex>
+#include <sstream>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+/**
+ * A C++ wrapper for Google's Service Account Credentials.
+ *
+ * Takes a JSON object representing the contents of a service account keyfile,
+ * and uses Google's OAuth2 service to obtain an access token.
+ *
+ * @warning
+ * The current implementation is a placeholder to unblock development of the
+ * Google Cloud Storage client libraries. There is substantial work needed
+ * before this class is complete, in fact, we do not even have a complete set of
+ * requirements for it.
+ *
+ * @see
+ *   https://developers.google.com/identity/protocols/OAuth2ServiceAccount
+ *   https://tools.ietf.org/html/rfc7523
+ *
+ * @tparam HttpRequestType a dependency injection point to make HTTP requests.
+ */
+template <typename HttpRequestType = CurlRequest,
+          typename ClockType = std::chrono::system_clock>
+class ServiceAccountCredentials : public storage::Credentials {
+ public:
+  explicit ServiceAccountCredentials(std::string const& content)
+      : ServiceAccountCredentials(content, GoogleOAuthRefreshEndpoint()) {}
+
+  explicit ServiceAccountCredentials(std::string const& content,
+                                     std::string oauth_server)
+      : requestor_(std::move(oauth_server)), expiration_time_(), clock_() {
+    auto credentials = nl::json::parse(content);
+    // This is the value of grant_type for:
+    // - JSON-formatted service account keyfiles downloaded from Cloud Console
+    nl::json assertion_header = {
+        {"alg", "RS256"},
+        {"kid", credentials["private_key_id"].get_ref<std::string const&>()},
+        {"typ", "JWT"}};
+
+    // TODO(#770): Remove all scopes except "cloud-platform".
+    std::ostringstream scope_oss;
+    scope_oss << GoogleOAuthScopeCloudPlatform() << " "
+              << GoogleOAuthScopeCloudPlatformReadOnly() << " "
+              << GoogleOAuthScopeDevstorageFullControl() << " "
+              << GoogleOAuthScopeDevstorageReadOnly() << " "
+              << GoogleOAuthScopeDevstorageReadWrite();
+    // Some credential formats (e.g. gcloud's ADC file) don't contain a
+    // "token_uri" attribute in the JSON object.  In this case, we try using the
+    // default value. See the comments around GoogleOAuthRefreshEndpoint about
+    // potential drawbacks to this approach.
+    std::string token_uri;
+    const char TOKEN_URI_KEY[] = "token_uri";
+    if (credentials.count(TOKEN_URI_KEY)) {
```

Comment:
I have since learned that you can do this more easily with:

```C++
std::string token_uri = credentials.value("token_uri", GoogleOauthRefreshEndpoint());
```

http://nlohmann.github.io/json/

Reply:
Done.